### DISCUSSION TITLE
Why use standard LayerNorm over RMSNorm in JAX/Flax? Does XLA fusion negate the algorithmic disadvantage?

### DISCUSSION THREAD

**Discussion by AnirudhSKrishnan** [google/flax#5235] [Q&A]
I'm implementing normalization in Flax and trying to decide between standard LayerNorm and RMSNorm.

Usually RMSNorm is preferred because it's faster (skips the mean calculation/centering). But with jax.jit, doesn't XLA fuse the standard LayerNorm math into a single kernel anyway?

Does the "two-pass" nature of standard LayerNorm actually matter once it's compiled, or is the performance difference negligible compared to the memory bandwidth savings from fusion? Just trying to figure out if it's worth writing a custom RMSNorm or if I should just stick to the standard nn.LayerNorm.

Thanks.

**Comment by vfdev-5** [comment]
Flax NNX also provides [nnx.RMSNorm](https://github.com/google/flax/blob/a972a4dff0f2816789508a7a492d4d99cc2ab08e/flax/nnx/nn/normalization.py#L581)
